### ISSUE TITLE
DDP strategy does not wait for all predictions to complete

### ISSUE THREAD

**Issue by vikigenius** [Lightning-AI/pytorch-lightning#17521]
### Bug description

Here is the relevant codesnippet

```python
    strategy = "ddp" if devices > 1 and accelerator == "gpu" else "auto"
    trainer = pl.Trainer(
        logger=False,
        strategy=strategy,
        accelerator=accelerator,
        devices=devices,
        deterministic=True,
    )
    pred_batches = trainer.predict(model, datamodule=datamodule)
    predictions = [pred for preds in pred_batches for pred in preds["questions"]]  # type: ignore
    #import pdb
    #pdb.set_trace()
    output_path.write_text("predictions\n" + "\n".join(predictions))
```
In my example the data size is 597

If I use multiple GPU devices and DDP strategy then my output file contains only something like 74 predictions. However if I use only a single GPU and the "auto" strategy, I can get back 597 predictions in the output file.

I was under the impression that the code after the predict function is executed only in a single process?
```
    pred_batches = trainer.predict(model, datamodule=datamodule)
    predictions = [pred for preds in pred_batches for pred in preds["questions"]]  # When we reach here, have we moved out of the distributed setting?
```

### What version are you seeing the problem on?

v2_0

### How to reproduce the bug

```python
1. Pick any small dataset where total size is not divisible by batch size.
2. Use DDP strategy with multiple GPUs
3. Use the predict function to do inference.
4. Try writing data to file.
5. Program exists before all data is written to file.
```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
<details>
  <summary>Current environment</summary>

* CUDA:
	- GPU:
		- NVIDIA A100-SXM4-80GB
		- NVIDIA A100-SXM4-80GB
		- NVIDIA A100-SXM4-80GB
		- NVIDIA A100-SXM4-80GB
		- NVIDIA A100-SXM4-80GB
		- NVIDIA A100-SXM4-80GB
		- NVIDIA A100-SXM4-80GB
		- NVIDIA A100-SXM4-80GB
	- available:         True
	- version:           11.7
* Lightning:
	- lightning-utilities: 0.8.0
	- pytorch-lightning: 2.0.2
	- torch:             2.0.0
	- torch-model-archiver: 0.7.1
	- torchmetrics:      0.11.4
* Packages:
	- accelerate:        0.18.0
	- aim:               3.17.3
	- aim-ui:            3.17.3
	- aimrecords:        0.0.7
	- aimrocks:          0.4.0
	- aiofiles:          23.1.0
	- aiohttp:           3.8.4
	- aiosignal:         1.3.1
	- alembic:           1.10.4
	- anyio:             3.6.2
	- arger:             1.4.8
	- async-timeout:     4.0.2
	- attrs:             23.1.0
	- backoff:           2.2.1
	- base58:            2.0.1
	- boto3:             1.26.118
	- botocore:          1.29.118
	- cachetools:        5.3.0
	- cattrs:            22.2.0
	- certifi:           2022.12.7
	- cffi:              1.15.1
	- charset-normalizer: 3.1.0
	- click:             8.1.3
	- cmake:             3.26.3
	- coverage:          7.2.3
	- cryptography:      40.0.2
	- datasets:          2.11.0
	- dill:              0.3.6
	- docstring-to-markdown: 0.12
	- dparse:            0.6.2
	- enum-compat:       0.0.3
	- exceptiongroup:    1.1.1
	- execnet:           1.9.0
	- fastapi:           0.95.1
	- filelock:          3.12.0
	- frozenlist:        1.3.3
	- fsspec:            2023.4.0
	- gmpy2:             2.1.2
	- greenlet:          2.0.2
	- h11:               0.14.0
	- huggingface-hub:   0.14.0
	- idna:              3.4
	- iniconfig:         2.0.0
	- jedi:              0.18.2
	- jinja2:            3.1.2
	- jmespath:          1.0.1
	- joblib:            1.2.0
	- lightning-utilities: 0.8.0
	- lit:               16.0.2
	- lsprotocol:        2023.0.0a1
	- mako:              1.2.4
	- markdown-it-py:    2.2.0
	- markupsafe:        2.1.2
	- mdurl:             0.1.2
	- monotonic:         1.6
	- mpmath:            1.3.0
	- multidict:         6.0.4
	- multiprocess:      0.70.14
	- mypy:              1.2.0
	- mypy-extensions:   1.0.0
	- networkx:          3.1
	- nltk:              3.8.1
	- numpy:             1.24.3
	- nvidia-cublas-cu11: 11.10.3.66
	- nvidia-cuda-cupti-cu11: 11.7.101
	- nvidia-cuda-nvrtc-cu11: 11.7.99
	- nvidia-cuda-runtime-cu11: 11.7.99
	- nvidia-cudnn-cu11: 8.5.0.96
	- nvidia-cufft-cu11: 10.9.0.58
	- nvidia-curand-cu11: 10.2.10.91
	- nvidia-cusolver-cu11: 11.4.0.1
	- nvidia-cusparse-cu11: 11.7.4.91
	- nvidia-nccl-cu11:  2.14.3
	- nvidia-nvtx-cu11:  11.7.91
	- packaging:         21.3
	- pandas:            2.0.1
	- parso:             0.8.3
	- pillow:            9.5.0
	- pip:               23.0.1
	- pluggy:            1.0.0
	- pprintpp:          0.4.0
	- protobuf:          4.22.3
	- psutil:            5.9.5
	- py3nvml:           0.2.7
	- pyarrow:           11.0.0
	- pycparser:         2.21
	- pydantic:          1.10.7
	- pygls:             1.0.1
	- pygments:          2.15.1
	- pyparsing:         3.0.9
	- pytest:            7.3.1
	- pytest-clarity:    1.0.1
	- pytest-cov:        4.0.0
	- pytest-randomly:   3.12.0
	- pytest-sugar:      0.9.7
	- pytest-xdist:      3.2.1
	- python-dateutil:   2.8.2
	- python-lsp-jsonrpc: 1.0.0
	- python-lsp-server: 1.7.2
	- pytorch-lightning: 2.0.2
	- pytz:              2023.3
	- pyyaml:            6.0
	- regex:             2023.3.23
	- requests:          2.28.2
	- responses:         0.18.0
	- restrictedpython:  6.0
	- rich:              13.3.4
	- ruamel.yaml:       0.17.21
	- ruamel.yaml.clib:  0.2.7
	- ruff:              0.0.262
	- ruff-lsp:          0.0.24
	- s3transfer:        0.6.0
	- safety:            2.3.5
	- segment-analytics-python: 2.2.2
	- sentencepiece:     0.1.98
	- setuptools:        67.7.2
	- six:               1.16.0
	- sniffio:           1.3.0
	- sqgen:             2.0.0
	- sqlalchemy:        1.4.47
	- starlette:         0.26.1
	- sympy:             1.11.1
	- termcolor:         2.3.0
	- tokenizers:        0.13.3
	- toml:              0.10.2
	- tomli:             2.0.1
	- torch:             2.0.0
	- torch-model-archiver: 0.7.1
	- torchmetrics:      0.11.4
	- tqdm:              4.65.0
	- transformers:      4.28.1
	- triton:            2.0.0
	- typeguard:         2.13.3
	- typing-extensions: 4.5.0
	- tzdata:            2023.3
	- ujson:             5.7.0
	- urllib3:           1.26.15
	- uvicorn:           0.21.1
	- wheel:             0.40.0
	- xmltodict:         0.13.0
	- xxhash:            3.2.0
	- yarl:              1.9.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         
	- python:            3.10.11
	- version:           #1 SMP Debian 5.10.162-1 (2023-01-21)

</details>

```

</details>

### More info

The docs do not mention any synchronization steps to be necessary after calling the predict function. Is it needed?

**Comment by awaelchli** [comment]
Your code currently overwrites the prediction output file in each process (probably). You can try this to verify all your predictions are there:

```py
pred_batches = trainer.predict(model, datamodule=datamodule)
predictions = [pred for preds in pred_batches for pred in preds["questions"]]  # type: ignore

# make a file per rank, so each process can save its predictions into it
output_path = Path(f"out", "predictions-rank-{trainer.global_rank}.txt")
output_path.write_text("predictions\n" + "\n".join(predictions))
```

It can also be done with a callback (which is currently the recommended way to do distributed inference in Lightning):
https://lightning.ai/docs/pytorch/stable/deploy/production_basic.html#enable-distributed-inference

**Comment by vikigenius** [comment]
@awaelchli is there some documentation on how the batches are distributed to different ranks. Assume that I don't shuffle the dataset for predictions. Does it mean the first batch will go to rank 0, 2nd batch to rank 2 etc.?

Because I see that there is a batch_indices argument is it a global batch index or is it local for each rank?

**Comment by awaelchli** [comment]
The batches in each process get different samples yes. The batch_indices should hold the list of which indices from the dataset have made it into the current batch being passed there. This can be useful if you later need to associate specific inputs in the dataset with the predictions that were produced.
